extensions/vscode/lsp-client/src/extension.ts

```diff
@@ -51,12 +46,316 @@ export function activate(context: ExtensionContext) {
+			TimelinePanel.currentPanel._panel.reveal(column);
+			return;
+		}
+
+		// Otherwise, create a new panel.
+		const panel = vscode.window.createWebviewPanel(
+			TimelinePanel.viewType,
+			'Timeline',
+			column || vscode.ViewColumn.One,
+			getWebviewOptions(),
+		);
+
+		TimelinePanel.currentPanel = new TimelinePanel(panel);
+	}
+
+	public static revive(panel: vscode.WebviewPanel) {
+		TimelinePanel.currentPanel = new TimelinePanel(panel);
+	}
+
+	private constructor(panel: vscode.WebviewPanel) {
+		this._panel = panel;
+
+		// Set the webview's initial html content
+		this._update();
+
+		// Listen for when the panel is disposed
+		// This happens when the user closes the panel or when the panel is closed programmatically
+		this._panel.onDidDispose(() => this.dispose(), null, this._disposables);
+
+		// Update the content based on view changes
+		this._panel.onDidChangeViewState(
+			e => {
+				if (this._panel.visible) {
+					this._update();
+				}
+			},
+			null,
+			this._disposables
+		);
+
+		// Handle messages from the webview
+		this._panel.webview.onDidReceiveMessage(
+			message => {
+				switch (message.command) {
+					case 'alert':
+						vscode.window.showErrorMessage(message.text);
+						return;
+				}
+			},
+			null,
+			this._disposables
+		);
+	}
+
+	private _getDB() {
+		return new sqlite3.Database(__dirname + '/../../../../wake.db', sqlite3.OPEN_READONLY, (err) => {
+			if (err) {
+				return console.error(err.message);
+			}
+		});
+	}
+
+	public refresh() {
+		let db = this._getDB();
+		this._queryDB(db, (jobsJson: JobReflection[], accessesJson: { type: number, job: number }[]) => {
+			let message = {
+				jobReflections: jobsJson,
+				fileAccesses: accessesJson
+			}
+			// Send a message to the webview.
+			this._panel.webview.postMessage(message);
+			db.close();
+		});
+	}
+
+	public dispose() {
+		TimelinePanel.currentPanel = undefined;
+
+		// Clean up our resources
+		this._panel.dispose();
+
+		while (this._disposables.length) {
+			const x = this._disposables.pop();
+			if (x) {
+				x.dispose();
+			}
+		}
+	}
+
+	private _update() {
+		this._panel.title = 'Timeline';
+
+		let db = this._getDB();
+		try {
+			let data = fs.readFileSync(__dirname + '/../../../../share/wake/html/timeline.html', { encoding: 'utf8' }).toString();
+			this._queryDB(db, (jobsJson: JobReflection[], accessesJson: { type: number, job: number }[]) => {
+				this._panel.webview.html = this._formatString(data, JSON.stringify(jobsJson), JSON.stringify(accessesJson));
+				db.close();
+			});
+		} catch (err) { }
+		return;
+	}
+
+	private _formatString(s: string, ...args: string[]): string {
+		return s.replace(/{(\d+)}/g, function (match, number) {
+			return typeof args[number] != 'undefined'
+				? args[number] : match;
+		});
+	}
+
+	private _queryDB(db: sqlite3.Database, callback: Function) {
+		let jobsJson: JobReflection[] = [];
+		let accessesJson: { type: number, job: number }[] = [];
+		let jobsSql =
```

Comment:
I think I would format this differently personally so that its a bit more compact but thats just my personal preference.